pkg/storer/reserve.go

```diff
@@ -172,69 +172,71 @@ func (db *DB) evictionWorker(ctx context.Context) {
-		if err != nil {
-			db.logger.Error(err, "get expired batches")
-			return
-		}
-
-		if len(batchesToEvict) == 0 {
-			return
-		}
+		go func() {
+			defer expirySem.Release(1)
 
-		// we ensure unreserve is not running and if it is we cancel it and wait
-		// for it to finish, this is to prevent unreserve and expirations running
-		// at the same time. The expiration will free up space so the unreserve
-		// target might change. This is to prevent unreserve from running with
-		// the old target.
-		if !unreserveSem.TryAcquire(1) {
-			cancelUnreserve()
-			err := unreserveSem.Acquire(ctx, 1)
+			batchesToEvict, err := db.getExpiredBatches()
 			if err != nil {
-				db.logger.Error(err, "acquire unreserve semaphore")
+				db.logger.Error(err, "get expired batches")
 				return
 			}
-			// trigger it again at the end if required
-			defer db.events.Trigger(reserveOverCapacity)
-		}
-		defer unreserveSem.Release(1)
-
-		defer db.events.Trigger(reserveUnreserved)
 
-		db.metrics.ExpiryRunsCount.Inc()
-
-		for _, batchID := range batchesToEvict {
-			b := batchID
-			if err := expiryWorkers.Acquire(ctx, 1); err != nil {
-				db.logger.Error(err, "acquire expiry worker semaphore")
+			if len(batchesToEvict) == 0 {
 				return
 			}
-			go func() {
-				defer expiryWorkers.Release(1)
 
-				err := db.removeExpiredBatch(ctx, b)
+			// we ensure unreserve is not running and if it is we cancel it and wait
+			// for it to finish, this is to prevent unreserve and expirations running
+			// at the same time. The expiration will free up space so the unreserve
+			// target might change. This is to prevent unreserve from running with
+			// the old target.
+			if !unreserveSem.TryAcquire(1) {
+				cancelUnreserve()
+				err := unreserveSem.Acquire(ctx, 1)
 				if err != nil {
-					db.logger.Error(err, "remove expired batch", "batch_id", hex.EncodeToString(b))
-				} else {
-					db.metrics.ExpiredBatchCount.Inc()
+					db.logger.Error(err, "acquire unreserve semaphore")
+					return
 				}
-			}()
-		}
+				// trigger it again at the end if required
+				defer db.events.Trigger(reserveOverCapacity)
+			}
+			defer unreserveSem.Release(1)
 
-		if err := expiryWorkers.Acquire(ctx, 4); err != nil {
-			db.logger.Error(err, "wait for expiry workers")
-			return
-		}
-		expiryWorkers.Release(4)
+			defer db.events.Trigger(reserveUnreserved)
+
+			db.metrics.ExpiryRunsCount.Inc()
+
+			for _, batchID := range batchesToEvict {
+				b := batchID
+				if err := expiryWorkers.Acquire(ctx, 1); err != nil {
+					db.logger.Error(err, "acquire expiry worker semaphore")
+					return
+				}
+				go func() {
+					defer expiryWorkers.Release(1)
+
+					err := db.removeExpiredBatch(ctx, b)
+					if err != nil {
+						db.logger.Error(err, "remove expired batch", "batch_id", hex.EncodeToString(b))
+					} else {
+						db.metrics.ExpiredBatchCount.Inc()
+					}
+				}()
+			}
+
+			if err := expiryWorkers.Acquire(ctx, 4); err != nil {
```

Comment:
what is the purpose of this?

Reply:
This is to ensure all expiry workers finished before we give up the sem. At the most, we will start 4 workers in parallel to expire. There are more workers here as for expired batches each of them will be working on different entries, so they can all do it in parallel.

Reply:
i see, we can have at most 4 `removeExpiredBatch` funcs running in parallel. the `Acquire` here waits for the above calls to finish.